src/lib/components/forms/Projects.svelte

```diff
@@ -0,0 +1,162 @@
+  import { page } from "$app/stores";
+
+  import { createEventDispatcher, getContext, onMount } from "svelte";
+  import { _ } from "svelte-i18n";
+
+  import Button from "../common/Button.svelte";
+  import Field from "../inputs/Field.svelte";
+  import Flex from "../common/Flex.svelte";
+  import Switch from "../inputs/Switch.svelte";
+  import Text from "../inputs/Text.svelte";
+  import TextArea from "../inputs/TextArea.svelte";
+
+  import { getForUser, add, remove } from "$lib/api/projects";
+  import { getCsrfToken } from "$lib/utils/api";
+  import { intersection } from "@/util/array.js";
+
+  export let documents: Document[] = [];
+  export let projects: Project[] = [];
+
+  const dispatch = createEventDispatcher();
+  const me: Writable<User> = getContext("me");
+
+  let common: Set<number>;
+
+  $: common = new Set(
+    intersection(
+      documents.map((d) => d.projects),
+      (a, b) => {
+        return a.id === b.id;
+      },
+    ).map((p: Project | number) => (typeof p === "number" ? p : p.id)),
+  );
+
+  $: console.log(projects);
+
+  onMount(async () => {
+    if ($me && projects.length === 0) {
+      projects = await getForUser($me.id);
+    }
+  });
+
+  function invalidateDocs(documents: Document[]) {
+    return Promise.all(documents.map((d) => invalidate(`document:${d.id}`)));
+  }
+
+  // typescript doesn't know what to do with svelte's events
+  async function toggle(project: Project, e) {
+    const { checked } = e.target;
+    const ids = documents.map((d) => d.id);
+    const csrf_token = getCsrfToken();
+    if (checked) {
+      await add(project.id, ids, csrf_token);
+    } else {
+      await remove(project.id, ids, csrf_token);
+    }
+    await invalidateDocs(documents);
+  }
+
+  function onSubmit({ formElement, formData, action, cancel, submitter }) {
+    submitter.disabled = true;
+    return async ({ result, update }) => {
+      if (result.type === "success") {
+        projects = [...projects, result.data.project];
+        update(result);
+      }
+      submitter.disabled = false;
+    };
+  }
+</script>
+
+<div class="container">
+  <form method="post" action="/documents/projects/" use:enhance={onSubmit}>
+    <h2>{$_("projects.create")}</h2>
+    <Field title={$_("projects.fields.title")} required inline>
+      <Text name="title" placeholder={$_("projects.fields.title")} />
+    </Field>
+
+    <details>
+      <summary>{$_("common.more")} &hellip;</summary>
+      <Flex direction="column">
+        <Field title={$_("projects.fields.description")}>
+          <TextArea name="description" />
+        </Field>
+        <Field title={$_("projects.fields.private")} inline>
+          <Switch name="private" />
+        </Field>
+        <Field title={$_("projects.fields.pinned")} inline>
+          <Switch name="pinned" />
+        </Field>
+      </Flex>
+    </details>
+    <Flex>
+      <Button mode="primary" type="submit">
+        {$_("common.add")} +
+      </Button>
+    </Flex>
+  </form>
```

Comment:
I think that this form belongs in its own modal:

- Creating a new project is a focused and intentional interaction that waits on a server response, so it creates a new UI context. This points to a modal being the correct space for this interaction.
- Creating a new project is a less common action when organizing documents. In this UI, we can assume the frequency of a user creating new projects is less than the frequency of picking existing projects.
- Use of a `<details>` element speaks to the current UI feeling too "cluttered" with the full creation form, but I think this misplaces the disclosure of additional interaction. Instead of hiding individual inputs from users, we should hide the entire form until the user makes an intentional interaction to create a new project.
- This project form has no opinion about the context it's rendered in, so it could appear in any kind of layout. This means we are not necessarily stacking modals inside one interface. Our current implementation of modals also supports layering, so the question is not "will this decision result in a nested modal?" but "is a modal the appropriate interaction in the context of this component?" In this case, the answer is yes. By reframing the question, it allows us to think more critically about the use of a modal in the context of each component—the culmination of those decisions is a consisent and coherent UI for the user.

Reply:
I think you're right, but I'm going to punt this to a separate issue/PR in favor of getting this merged sooner. I'll open an issue to refactor.